nixos/modules/services/mail/listmonk.nix

```diff
@@ -0,0 +1,223 @@
+        };
+        mutableSettings = mkOption {
+          type = types.bool;
+          default = true;
+          description = lib.mdDoc ''
+            Database settings will be reset to the value set in this module if this is not enabled.
+            Enable this if you want to persist changes you have done in the application.
+          '';
+        };
+      };
+      package = mkPackageOption pkgs "listmonk" {};
+      settings = mkOption {
+        type = types.submodule { freeformType = tomlFormat.type; };
+        description = lib.mdDoc ''
+          Static settings set in the config.toml, see <https://github.com/knadh/listmonk/blob/master/config.toml.sample> for details.
+          You can set secrets using the secretFile option with environment variables following <https://listmonk.app/docs/configuration/#environment-variables>.
+        '';
+      };
+      secretFile = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = lib.mdDoc
+          "A file containing secrets as environment variables. See <https://listmonk.app/docs/configuration/#environment-variables> for details on supported values.";
+      };
+    };
+  };
+
+  ###### implementation
+  config = mkIf cfg.enable {
+    # Default parameters from https://github.com/knadh/listmonk/blob/master/config.toml.sample
+    services.listmonk.settings."app".address = mkDefault "localhost:9000";
+    services.listmonk.settings."db" = mkMerge [
+      ({
+        max_open = mkDefault 25;
+        max_idle = mkDefault 25;
+        max_lifetime = mkDefault "300s";
+      })
+      (mkIf cfg.database.createLocally {
+        host = mkDefault "/run/postgresql";
+        port = mkDefault 5432;
+        user = mkDefault "listmonk";
+        database = mkDefault "listmonk";
+      })
+    ];
+
+    services.postgresql = mkIf cfg.database.createLocally {
+      enable = true;
+
+      ensureUsers = [{
+        name = "listmonk";
+        ensurePermissions = { "DATABASE listmonk" = "ALL PRIVILEGES"; };
+      }];
+
+      ensureDatabases = [ "listmonk" ];
+    };
+
+    systemd.services.listmonk = {
+      description = "Listmonk - newsletter and mailing list manager";
+      after = [ "network.target" ]
+        ++ optional cfg.database.createLocally "postgresql.service";
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "exec";
+        EnvironmentFile = mkIf (cfg.secretFile != null) [ cfg.secretFile ];
+        ExecStartPre = [
+          # StateDirectory cannot be used when DynamicUser = true is set this way.
+          # Indeed, it will try to create all the folders and realize one of them already exist.
+          # Therefore, we have to create it ourselves.
+          # $STATE_DIRECTORY is not available at this step.
+          ''${pkgs.coreutils}/bin/mkdir -p "''${STATE_DIRECTORY}/listmonk/uploads"''
```

Comment:
Better written as `serviceConfig.StateDirectory = [ "listmonk" "listmonk/uploads" ];`.

Reply:
It is not possible alas, @aanderse ; as stated in the comments above (I should remove the `$STATE_DIRECTORY` comment).
Adding this path will fail as it will remark that `/var/lib/listmonk` is already created.

Reply:
Sorry, I just didn't (still don't) understand the comment :man_shrugging:

Reply:
@aanderse No problem, if I do use `DynamicUser = true;` with multiple entries such that prefixes of some are in the array, here's the error:
```
machine # [    8.743574] systemd[945]: listmonk.service: Failed to set up special execution directory in /var/lib: File exists
machine # [    8.745210] nscd[880]: 880 monitoring directory `/etc` (2)
machine # [    8.745777] systemd[945]: listmonk.service: Failed at step STATE_DIRECTORY spawning /nix/store/bii7l73s4mldagsfrn86161xzsns1810-listmonk-2.2.0/bin/listmonk: File exists
machine # [    8.747025] nscd[880]: 880 monitoring file `/etc/passwd` (1)
machine # [    8.747586] systemd[1]: listmonk.service: Control process exited, code=exited, status=238/STATE_DIRECTORY
```

@mweinelt told me that he thinks that DynamicUser + multiple StateDirectories this way is not supported (or is this a bug?) (let me know if I'm wrong).

Therefore, I have to create it in the `ExecPreStart` manually.

Reply:
Very strange output. Definitely sounds like a bug in `systemd`! Please consider creating a minimal example and reporting upstream.

Thanks for the explanation!

Reply:
The problem here is that systemd is irritated by both paths sharing a prefix. For the second directory it is irritated by the fact that the directory already exists. IMO a bug.

Reply:
see https://github.com/systemd/systemd/issues/24783 for upstream bug.

Reply:
Thanks!